### PR TITLE
stampy updates stamp counts in coda users table

### LINKED ISSUE #209: Sync stamp count to users table in coda
On stamp event: go to [coda's users table](https://coda.io/d/AI-Safety-Info_dfau7sl2hmG/Team_sur3i#Team_tu_Rc/r5), update the score of sender and recipient if there is a matching Discord handle. Once per ~day update all rows.

It's OK that users need to add themselves to the table, opt-in leaderboard is fine.

### PR DESCRIPTION
**Pull request by MatthewBaggins**
closes #209

### CODE REVIEW

**Comment by MatthewBaggins** [comment]
Whenever somebody's message is (un)stamped, Stampy updates that person's stamp count in [the table](https://coda.io/d/AI-Safety-Info_dfau7sl2hmG/Team_sur3i#Team_tu_Rc/r5). 

There's also update of all users' stamps, which gets executed on boot. The boot time is saved in `StampsModule.last_total_stamp_update`. When somebody's message is (un)stamped, Stampy checks that time and if it was earlier than 23 hours ago, he updates all people's stamp counts and saves the date again.

I also moved some coda-related methods to utilities, I'll probably move them the rest of them over the week.

File: modules/stampcollection.py [hunk 1]
```diff
@@ -159,9 +212,11 @@ def load_votes_from_csv(self, filename="stamps.csv"):
     async def load_votes_from_history(self):
         """Load up every time any stamp has been awarded by anyone in the whole history of the Discord
         This is omega slow, should basically only need to be called once"""
-        guild = discord.utils.find(lambda g: g.name == self.utils.GUILD, self.utils.client.guilds)
+        guild = discord.utils.find(
+            lambda g: g.name == self.utils.GUILD, self.utils.client.guilds
+        )
 
-        with open("stamps.csv", "w") as stamplog:
+        with open("stamps.csv", "w", encoding="utf-8") as stamplog:
```
Comment (Aprillion) [inline modules/stampcollection.py:219]: huh, utf-8 is not used by default?

File: modules/stampcollection.py [hunk 2]
```diff
@@ -58,6 +74,38 @@ def update_vote(self, emoji: str, from_id: int, to_id: int,
...
+    def update_stamps_in_users_table(
+        self, user: DiscordUser, stamp_count: float
+    ) -> None:
+        """Update stamps count in Coda [users/team table](https://coda.io/d/AI-Safety-Info_dfau7sl2hmG/Team_sur3i#_lu_Rc)"""
+        query = f'"Discord handle":"{get_user_handle(user)}"'
+        user_row = utils.get_user_row(query)
+        if user_row is None:
+            return
+        user_id = user_row["id"]
+        uri = f"https://coda.io/apis/v1/docs/{utils.DOC_ID}/tables/{utils.TEAM_GRID_ID}/rows/{user_id}"
```
Comment (Aprillion) [inline modules/stampcollection.py:86]: hm, production Coda tables should be probably updated only from production Discord server and not from the test server...

please discuss with @plexish whether we can use different columns (or different tables or whole different Coda docs) for testing

**Review by Aprillion — COMMENTED** [review]
same files are changes as in #213, so leaving up to @tayler6000 to discuss order of merge(-conflicts)

File: modules/stampcollection.py (hunk 2 above)
Comment (plexish) [reply modules/stampcollection.py:86]: A test copy of the whole system now exists: https://coda.io/d/AI-Test-Info_dah62XEPvpG/Admin-Panel_su93i?playModeWorkflowId=#_luBHo

**Review by Lovkush-A — COMMENTED** [review]
had a look. mostly just aesthetic suggestions below. how can i check if the new functionality works as expected?

File: utilities/__init__.py [hunk 3]
```diff
@@ -1,12 +1,14 @@
 from utilities import utilities as util
```
[the remark this answers is not in the document]
Comment (MatthewBaggins) [reply utilities/__init__.py:1]: we don't, thanks

File: utilities/utilities.py [hunk 4]
```diff
@@ -506,6 +511,47 @@ def split_message_for_discord(msg: str, stop_char: str = "\n", max_length: int =
...
 
+    ##############
+    # Coda utils #
+    ##############
+    
+    def get_coda_auth_headers(self) -> dict:
+        """Get authorization headers for coda requests"""
+        return {"Authorization": f"Bearer {self.CODA_API_TOKEN}"}
+    
+    def get_user_row(self, query: str) -> Optional[dict]:
```
[the remark this answers is not in the document]
Comment (MatthewBaggins) [reply utilities/utilities.py:522]: I think these unions don't work in python3.9, do they?

**Comment by MatthewBaggins** [comment]
> had a look. mostly just aesthetic suggestions below. how can i check if the new functionality works as expected?

Thanks!
To test it, run this branch on test server, start (un)stamping people's messages on test server and see if their stampcounts change in [the table on AI Test Info](https://coda.io/d/AI-Test-Info_dah62XEPvpG/Team_sur3i#_luI23)

File: utilities/utilities.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Lovkush-A) [reply utilities/utilities.py:522]: yes you're correct. https://peps.python.org/pep-0604/

File: modules/stampcollection.py [hunk 5]
```diff
@@ -58,6 +74,38 @@ def update_vote(self, emoji: str, from_id: int, to_id: int,
...
+                "cells": [
+                    {"column": "Stamp count", "value": stamp_count},
+                ],
+            },
+        }
+        response = requests.put(
+            uri, headers=utils.get_coda_auth_headers(), json=payload, timeout=20
+        )
+        response.raise_for_status()
+        log.info("Updated %s's stamp count to %f", user.name, stamp_count)
```
[the remark this answers is not in the document]
Comment (MatthewBaggins) [reply modules/stampcollection.py:98]: you can but it is generally advised to format strings with `%` in logging for performance reasons
doesn't matter really unless you do a lot of logging but whatever

**Comment by Lovkush-A** [comment]
Run tests.
- adding/removing stamps to rob's comments changed his stamp score on coda. fantastic!
- i think need to update logging. will make a code suggestion. check if other logging introduced here needs to be updated too

File: modules/stampcollection.py (hunk 5 above)
Comment (Lovkush-A) [inline modules/stampcollection.py:98]:
```suggestion
        log.info(self.class_name, "Updated %s's stamp count to %f on coda", user.name, stamp_count)
```

Comment (MatthewBaggins) [reply modules/stampcollection.py:98]: I don't this (putting `self.class_name` before the main string) is going to work, unless `structlog`'s logging works differently than standard python's logging. What should it output?

Comment (Lovkush-A) [reply modules/stampcollection.py]: i'm happy with your latest update to logging. it is different to how other ppl have done logging. i am new to all this so have no strong opinions, but what you did does look a bit cleaner:

image

File: utilities/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lovkush-A) [reply utilities/__init__.py:1]: I think top line of init is still there?

File: modules/stampcollection.py (hunk 5 above)
Comment (MatthewBaggins) [reply modules/stampcollection.py]: oh, wait, I just checked how most logging is done in this codebase and what you suggested was correct
yeah, it looks cleaner but I'll probably revert to what you suggested, although AFAIR Rob suggested to change the way we do logging at some point

**Review by Lovkush-A — APPROVED** [review]
all looks good and new functionality seems to work as expected (at least in test environment).

File: modules/stampcollection.py (hunk 5 above)
Comment (Lovkush-A) [reply modules/stampcollection.py]: i think we need:

`log.info(self.class_name, msg="...",)`

**Pull request merged**